core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/recovery/RecoveryQueue.scala

```diff
@@ -0,0 +1,153 @@
+package edu.uci.ics.amber.engine.architecture.recovery
+
+import edu.uci.ics.amber.engine.architecture.logging.{
+  InMemDeterminant,
+  LinkChange,
+  ProcessControlMessage,
+  SenderActorChange,
+  StepDelta,
+  TimeStamp
+}
+import edu.uci.ics.amber.engine.architecture.logging.storage.DeterminantLogStorage.DeterminantLogReader
+import edu.uci.ics.amber.engine.architecture.worker.WorkerInternalQueue
+import edu.uci.ics.amber.engine.architecture.worker.WorkerInternalQueue.{
+  ControlElement,
+  EndMarker,
+  EndOfAllMarker,
+  InputTuple,
+  InternalQueueElement,
+  SenderChangeMarker
+}
+import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity
+import lbmq.LinkedBlockingMultiQueue
+
+import java.util.concurrent.LinkedBlockingQueue
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+class RecoveryQueue(logReader: DeterminantLogReader) {
```

Comment:
I suggest we make `RecoveryQueue` and `WorkerInternalQueue` to implement the same public interface

Reply:
will do it in another PR.